amr-wind/equation_systems/icns/source_terms/VelocityFreeAtmosphereForcing.cpp

```diff
@@ -0,0 +1,152 @@
+
+#include "AMReX_ParmParse.H"
+#include "AMReX_Gpu.H"
+#include "AMReX_Random.H"
+
+namespace amr_wind::pde::icns {
+
+VelocityFreeAtmosphereForcing::VelocityFreeAtmosphereForcing(const CFDSim& sim)
+    : m_time(sim.time())
+    , m_mesh(sim.mesh())
+    , m_velocity(sim.repo().get_field("velocity"))
+    , m_sim(sim)
+{
+    amrex::ParmParse pp_abl("ABL");
+    pp_abl.query("rans_1dprofile_file", m_1d_rans);
+    if (!m_1d_rans.empty()) {
+        std::ifstream ransfile(m_1d_rans, std::ios::in);
+        if (!ransfile.good()) {
+            amrex::Abort("Cannot find 1-D RANS profile file " + m_1d_rans);
+        }
+        amrex::Real value1, value2, value3, value4, value5;
+        while (ransfile >> value1 >> value2 >> value3 >> value4 >> value5) {
+            m_wind_heights.push_back(value1);
+            m_u_values.push_back(value2);
+            m_v_values.push_back(value3);
+            m_w_values.push_back(value4);
+        }
+    } else {
+        amrex::Abort("Cannot find 1-D RANS profile file " + m_1d_rans);
+    }
+    pp_abl.query("meso_sponge_start", m_meso_start);
+    pp_abl.query("meso_timescale", m_meso_timescale);
+    int num_wind_values = static_cast<int>(m_wind_heights.size());
+    m_wind_heights_d.resize(num_wind_values);
+    m_u_values_d.resize(num_wind_values);
+    m_v_values_d.resize(num_wind_values);
+    m_w_values_d.resize(num_wind_values);
+    amrex::Gpu::copy(
+        amrex::Gpu::hostToDevice, m_wind_heights.begin(), m_wind_heights.end(),
+        m_wind_heights_d.begin());
+    amrex::Gpu::copy(
+        amrex::Gpu::hostToDevice, m_u_values.begin(), m_u_values.end(),
+        m_u_values_d.begin());
+    amrex::Gpu::copy(
+        amrex::Gpu::hostToDevice, m_v_values.begin(), m_v_values.end(),
+        m_v_values_d.begin());
+    amrex::Gpu::copy(
+        amrex::Gpu::hostToDevice, m_w_values.begin(), m_w_values.end(),
+        m_w_values_d.begin());
+}
+
+VelocityFreeAtmosphereForcing::~VelocityFreeAtmosphereForcing() = default;
+
+void VelocityFreeAtmosphereForcing::operator()(
+    const int lev,
+    const amrex::MFIter& mfi,
+    const amrex::Box& bx,
+    const FieldState fstate,
+    const amrex::Array4<amrex::Real>& src_term) const
+{
+    const auto& geom = m_mesh.Geom(lev);
+    const auto& dx = geom.CellSizeArray();
+    const auto& prob_lo = geom.ProbLoArray();
+    const auto& prob_hi = geom.ProbHiArray();
+    const auto& velocity =
+        m_velocity.state(field_impl::dof_state(fstate))(lev).const_array(mfi);
+    const amrex::Real sponge_start = m_meso_start;
+    const amrex::Real meso_timescale = m_time.delta_t();
+    const auto vsize = m_wind_heights_d.size();
+    const auto* wind_heights_d = m_wind_heights_d.data();
+    const auto* u_values_d = m_u_values_d.data();
+    const auto* v_values_d = m_v_values_d.data();
+    const auto* w_values_d = m_w_values_d.data();
+    const bool has_terrain = this->m_sim.repo().field_exists("terrain_height");
+    if (has_terrain) {
```

Comment:
let's consolidate these 2 branches since they look so similar. I think the only difference is the calculation of z? I would move the `if has_terrain` into the parallel for. It's not ideal but I would rather than then the code duplication.

Reply:
Or do it like you do below with the perturbation forcing